pubtools/_content_gateway/push_base.py

```diff
@@ -480,33 +534,140 @@ def process_file(self, file_item):
+            self.completed_operations.append(file_item)
+            return
+
+        if file_item.get('state') == 'update' and file_id:
+            LOG.info("Updating the file record \n")
             file_item.get('metadata')['id'] = file_id
             self.cgw_client.update_file(product_id, version_id, file_item.get('metadata'))
-        elif file_item.get('state') == 'absent' and file_id:
+
+            # adding operational product_id for rollback operation
+            file_item['product_id'] = product_id
+            # keeping a copy of data that has been used to update the cgw data
+            self.completed_operations.append(file_item)
+
+        elif file_item.get("state") == "delete" and file_id:
             LOG.info("Deleting existing file records for file_id:- %s" % file_id)
             self.cgw_client.delete_file(product_id, version_id, file_id)
-            LOG.info("File record deleted!")
+
+            #  removing the deleted file record from the completed operations
+            #  this is to avoid the conflicts during the rollback and make_visible operations
+            for record in self.completed_operations:
+                if record['type'] == 'file' and record['file_id'] == file_id:
+                    self.completed_operations.remove(record)
+
+            LOG.info("File record deleted! \n")
             del self.file_mapping[(product_name, product_code, version_name, download_url)]
         else:
-            raise CGWError("Cannot delete file: product version: '%s' product code: '%s' product: '%s', id is not set" %
-                           (version_name, product_code, product_name))
+            raise CGWError(
+                "Cannot update/delete file: product version: '%s' product code: '%s' product: '%s', id is not set" %
+                (version_name, product_code, product_name))
         return file_id
+
+    def make_visible(self):
+        """
+        - Update the invisible attribute to False
+          once all the CGW operation completed successfully.
+
+        - This is needed to make the version and file records visible.
+        - The `invisible` attribute changes on the following conditions:
+
+            - `invisible=True` :- sets if user externally specify the `invisible=True`
+            - `invisible=False` :- sets either if user externally specify the `invisible=False`
+              or user doesn't specify any value for this `invisible` attribute.
+
+        """
+
+        LOG.debug("Updating the invisible attribute of the created records")
+        LOG.debug("Enabling the created records if required \n")
+
+        self.completed_operations.reverse()
+        for item in self.completed_operations:
+            if item.get('type') == 'product_version':
+                if item.get("state") == "create":
+                    item["metadata"]["id"] = item["version_id"]
+                    item["metadata"]['invisible'] = False if not item["metadata"].get('invisible') else True
```

Comment:
isn't easier to just write:
item["metadata"].get('invisible') or False?

Reply:
Yes! will update it.